Fix #6672: Translation prioritization

shavavo opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

TODO: #9674 for @shavavo after merge

1. This PR fixes or fixes part of #6672
2. This PR does the following: [Design Doc](https://docs.google.com/document/d/1JdDG4NdVLS4PCm9KmQ-v3OwYDVmStFennAarkSdYB3A/edit?usp=sharing)
* Allows admin to specify featured translation languages in the Admin > Config tab
![Screen Shot 2020-06-05 at 12 24 20 PM](https://user-images.githubusercontent.com/33111844/83929502-90217700-a761-11ea-9903-dbe0fd023f04.png)

* Surfaces featured translation languages to contributors on the community dashboard in the translation tab
* Redesigned language selector on translation tab 

![Screen Shot 2020-06-05 at 5 26 45 PM](https://user-images.githubusercontent.com/33111844/83929540-b1826300-a761-11ea-982e-dfea0e164b14.png)


## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

Commit:
added config for featured language

Commit:
styling

Commit:
added get endpoint

Commit:
frontend backend service + tests

Commit:
styling; codeowners

Commit:
sort inputs

Commit:
sorted imports

Commit:
Styling

Commit:
formatting

Commit:
basic component setup

Commit:
fixed linting

Commit:
new selector completed

Commit:
styling

Commit:
EtE tests

Commit:
Added E2E test for switching languages

Commit:
e2e

Commit:
removed usused code

shavavo commented:
TODO @shavavo: Add tests for `core/templates/pages/community-dashboard-page/translation-language-select/translation-language-select.component.ts`

Update: done

U8NWXD commented on core/tests/protractor_utils/CommunityDashboardTranslateTextTab.js:
```diff
@@ -26,8 +26,12 @@ var CommunityDashboardTranslateTextTab = function() {
     by.css('option:checked'));
 
   var _selectLanguage = async function(language) {
+    await selectableLanguageElements.click();
```
Why do we need this? I think usually you can click directly on dropdown elements without clicking on the parent. I think your way is fine; I'm just trying to understand why it's different

shavavo replied:
The previous element was simply a <select /> component. This selector is a custom component, and the dropdown is set to display:none until it is clicked. Without the click, the next step would fail

U8NWXD replied:
makes sense!

U8NWXD reviewed (changes requested):
@shavavo Just a question

U8NWXD reviewed (approved):
lgtm

shavavo commented:
@sagangwee @DubeySandeep @marianazangrossi @nithusha21 PTAL

Commit:
tests are failing

Commit:
a few tests

sagangwee commented on core/domain/config_domain.py:
```diff
@@ -30,6 +30,24 @@
 
 CMD_CHANGE_PROPERTY_VALUE = 'change_property_value'
 
+LIST_OF_FEATURED_TRANSLATION_LANGUAGES_DICTS = {
```
Append `_SCHEMA` to be consistent with the other schemas?

shavavo replied:
Done

Commit:
finished tests

Commit:
relative reference exception

sagangwee commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -95,6 +103,34 @@ export class ContributionOpportunitiesBackendApiService {
+  ) {
+    this.http.get('/getfeaturedtranslationlanguages').toPromise()
+      .then((data: any) => {
+        let featuredTranslationLanguages = (
```
We should be able to use `const` here.

shavavo replied:
Done

Commit:
relative reference exception

Commit:
relative reference exception

Commit:
linting

Commit:
linting

Commit:
linting

sagangwee commented on core/templates/domain/community_dashboard/ReadOnlyFeaturedTranslationLanguageObjectFactory.ts:
```diff
@@ -0,0 +1,67 @@
+}
+
+export class ReadOnlyFeaturedTranslationLanguage {
+  _languageCode: string;
+  _description: string;
```
Prefer declaring as private rather than using leading underscores. Here and elsewhere. You can also declare these variables as `readonly` here since the only time they're being assigned is in the constructor.

shavavo replied:
Done

sagangwee commented on core/templates/domain/community_dashboard/ReadOnlyFeaturedTranslationLanguageObjectFactorySpec.ts:
```diff
@@ -0,0 +1,45 @@
+describe('Read only Featured Translation Language factory', () => {
+  let readOnlyFTLFactory:
+    ReadOnlyFeaturedTranslationLanguageObjectFactory = null;
+  var _sampleFTL = null;
```
`var` in JavaScript is function scoped, which can cause difficult to understand bugs, so prefer `let` (block scoped).

shavavo replied:
Done

sagangwee commented on core/templates/domain/community_dashboard/ReadOnlyFeaturedTranslationLanguageObjectFactory.ts:
```diff
@@ -0,0 +1,67 @@
+    this._description = _description;
+  }
+
+  getLanguageCode(): string {
+    return this._languageCode;
+  }
+
+  getDescription(): string {
+    return this._description;
+  }
```
You can convert these to TypeScript getters: https://www.typescriptlang.org/docs/handbook/classes.html#accessors

shavavo replied:
Done

sagangwee commented on core/templates/domain/community_dashboard/ReadOnlyFeaturedTranslationLanguageObjectFactory.ts:
```diff
@@ -0,0 +1,67 @@
+  _languageCode: string;
+  _description: string;
+
+  constructor(
+      _languageCode: string,
+      _description: string
+  ) {
+    this._languageCode = _languageCode;
+    this._description = _description;
+  }
```
You can use TypeScript's parameter properties here: https://www.typescriptlang.org/docs/handbook/classes.html#parameter-properties

shavavo replied:
Done

sagangwee commented on core/templates/domain/community_dashboard/ReadOnlyFeaturedTranslationLanguageObjectFactory.ts:
```diff
@@ -0,0 +1,67 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
Is it necessary to prepend this class/file with 'ReadOnly'? There doesn't seem to be a non-ReadOnly equivalent, so it doesn't seem to offer much value here.

shavavo replied:
Done

sagangwee commented on core/templates/domain/community_dashboard/ReadOnlyFeaturedTranslationLanguageObjectFactory.ts:
```diff
@@ -0,0 +1,67 @@
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+export interface IFeaturedTranslationLanguageBackendDict {
```
The "I" prefix is unnecessary for interfaces: it's enough to be describe the interface's intended behavior via the name.

shavavo replied:
I believe all backend dicts have an I prefix currently - correct me if I'm wrong

sagangwee replied:
This pattern for prefixing interfaces with 'I' does seem common throughout the codebase, but TypeScript already expresses information in types, so generally names should not be decorated with information that is included in the type. Same reasoning for preferring using the `private` modifier over an underscore prefix. I think we should move away from the 'I' naming practice moving forward.

sagangwee commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -95,6 +103,34 @@ export class ContributionOpportunitiesBackendApiService {
+        featuredTranslationLanguages:
+        ReadOnlyFeaturedTranslationLanguage[]
+      ) => void,
+      errorCallback: (reason?: any) => void
```
The optional parameter in this callback might not be necessary here: https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#optional-parameters-in-callbacks

shavavo replied:
Done

sagangwee reviewed (changes requested):
@shavavo Took a pass. Thanks! A lot of cool Angular 8 features in here.

nithusha21 reviewed (approved):
LGTM for codeowner files....

Commit:
linter

Commit:
Addressed comments

shavavo commented:
Thanks @sagangwee, those are some cool typescript features I haven't seen before. I addressed all your comments, PTAL.

Also have you bumped into this? My linting tests are failing because of this error:
`Unable to open file: oppia/tmpcompiledjsFlfu7X/translation-language-select.component.spec.js
`

I've submitted an issue for now, just wondering if you've seen it.

shavavo commented:
Note: changes to `webpack-common.config.js`, `karma.conf.ts`, `package.json`,  and`general_purpose_linter.py` are from PR #9398 and have passed review. These changes are required to have angular2+ component tests working

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

sagangwee commented on core/templates/domain/community_dashboard/FeaturedTranslationLanguageObjectFactorySpec.ts:
```diff
@@ -0,0 +1,43 @@
+  './FeaturedTranslationLanguageObjectFactory';
+
+describe('Featured Translation Language object factory', () => {
+  let readOnlyFTLFactory: FeaturedTranslationLanguageObjectFactory = null;
```
Update the name to reflect the dropping of the 'readOnly' prefix.

shavavo replied:
Done

sagangwee commented on core/templates/domain/community_dashboard/FeaturedTranslationLanguageObjectFactorySpec.ts:
```diff
@@ -0,0 +1,43 @@
+
+describe('Featured Translation Language object factory', () => {
+  let readOnlyFTLFactory: FeaturedTranslationLanguageObjectFactory = null;
+  let _sampleFTL = null;
```
The underscore prefix seems unnecessary here.

shavavo replied:
Done

sagangwee commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.spec.ts:
```diff
@@ -176,4 +180,36 @@ describe('Contribution Opportunities backend API service', function() {
+      const successHandler = jasmine.createSpy('success');
+      const failHandler = jasmine.createSpy('fail');
+
+      const readOnlyFTLFactory = TestBed.get(
```
Drop 'readOnly' here as well.

shavavo replied:
Done

sagangwee commented on core/templates/domain/community_dashboard/FeaturedTranslationLanguageObjectFactorySpec.ts:
```diff
@@ -0,0 +1,43 @@
+  beforeEach(() => {
+    readOnlyFTLFactory = TestBed.get(FeaturedTranslationLanguageObjectFactory);
+
+    let sampleFRLDict = {
```
What is FRL?

shavavo replied:
Typo - FTL. Done

sagangwee commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -46,7 +52,9 @@ export class ContributionOpportunitiesBackendApiService {
     private urlInterpolationService: UrlInterpolationService,
-    private http: HttpClient
+    private http: HttpClient,
+    private readOnlyFeaturedTranslationLanguageObjectFactory:
```
Drop 'readOnly'.

shavavo replied:
Done

sagangwee commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -95,6 +103,33 @@ export class ContributionOpportunitiesBackendApiService {
     });
   }
 
+  private _fetchFeaturedTranslationLanguages(
```
If declaring the method as private, then the underscore prefix becomes redundant. I think you'll have to resolve the resulting naming conflict though.

shavavo replied:
Done

sagangwee commented on core/templates/pages/community-dashboard-page/translation-language-select/translation-language-select.component.ts:
```diff
@@ -0,0 +1,99 @@
+
+  _hideDescriptionPopup() {
+    this.descriptionPopupShown = false;
+  }
```
Declare as private if you're using underscores? Though if you're accessing the methods in the templates, then I don't think these methods should be private anyway.

sagangwee commented on core/templates/pages/community-dashboard-page/translation-language-select/translation-language-select.component.spec.ts:
```diff
@@ -0,0 +1,206 @@
+  let component: TranslationLanguageSelectComponent;
+  let fixture: ComponentFixture<TranslationLanguageSelectComponent>;
+
+  let readOnlyFeaturedTranslationLanguageObjectFactory =
```
Update the 'readOnly' instances in this file as well.

shavavo replied:
Done

sagangwee commented on core/templates/domain/community_dashboard/FeaturedTranslationLanguageObjectFactory.ts:
```diff
@@ -0,0 +1,60 @@
+
+export class FeaturedTranslationLanguage {
+  constructor(
+      readonly _languageCode: string,
+      readonly _description: string
```
Remove the underscores.

shavavo replied:
Done.

sagangwee commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -95,6 +103,33 @@ export class ContributionOpportunitiesBackendApiService {
+      successCallback: (
+        featuredTranslationLanguages: FeaturedTranslationLanguage[]
+      ) => void,
+      errorCallback: (reason: any) => void
```
We should try to avoid using `any` if possible ("TypeScript's any type is a super and subtype of all other types, and allows dereferencing all properties. As such, any is dangerous - it can mask severe programming errors, and its use undermines the value of having static types in the first place."). Maybe we can enforce reason to be a `string` here.

shavavo replied:
Done

sagangwee commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -75,7 +83,7 @@ export class ContributionOpportunitiesBackendApiService {
         opportunities?: Array<any>, nextCursor?: string, more?: boolean
         ) => void,
-      errorCallback: (reason?: any) => void
+      errorCallback: (reason: any) => void
```
Can reason be a `string` here?

shavavo replied:
Done

sagangwee commented on core/templates/domain/community_dashboard/FeaturedTranslationLanguageObjectFactory.ts:
```diff
@@ -0,0 +1,60 @@
+      readonly _description: string
+  ) {}
+
+  get languageCode(): string {
+    return this.languageCode;
+  }
+
+  get description(): string {
+    return this.description;
+  }
```
Actually, since there is no additional logic in these methods (they're just returning the variable values), the getters aren't necessary here. Consumers can just access the readonly variables directly.

shavavo replied:
Done

sagangwee commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -95,6 +103,33 @@ export class ContributionOpportunitiesBackendApiService {
+                .createFromBackendDict(backendDict)
+          ));
+
+        if (successCallback) {
+          successCallback(featuredTranslationLanguages);
+        }
+      },
+      (error) => {
+        if (errorCallback) {
+          errorCallback(error);
+        }
+      });
```
Is this necessary? Can't we just return `featuredTranslationLanguages` and `then()` will return a new Promise?

shavavo replied:
Not entirely sure what you mean, but I did clean up the code here

sagangwee reviewed (changes requested):
@shavavo Took another pass. Thanks!

Commit:
cleaned up code

Commit:
styling

shavavo commented:
@sagangwee PTAL

sagangwee commented on core/templates/pages/community-dashboard-page/translation-language-select/translation-language-select.component.ts:
```diff
@@ -0,0 +1,99 @@
+  descriptionPopupContent = '';
+
+  constructor(
+    private contributionOpportunitiesBackendApiService:
+      ContributionOpportunitiesBackendApiService
```
Does `contributionOpportunitiesBackendApiService` need to be a class member? Can't we just invoke `ContributionOpportunitiesBackendApiService .fetchFeaturedTranslationLanguages()` directly?

shavavo replied:
@sagangwee correct me if I'm wrong, but fetchFeaturedTranslationLanguages is not a static method. 

I think this is how we inject dependencies in Angular (https://angular.io/guide/architecture-services), I'm not sure I've seen an instance on directly calling a backend api service in the code base.

sagangwee replied:
Ah, gotcha. Didn't realize the service was migrated over to Angular 8.

sagangwee commented on core/templates/domain/community_dashboard/FeaturedTranslationLanguageObjectFactory.ts:
```diff
@@ -0,0 +1,52 @@
+export class FeaturedTranslationLanguageObjectFactory {
+  createFromBackendDict(
+      featuredTranslationBackendDict: FeaturedTranslationLanguageBackendDict
+  ):FeaturedTranslationLanguage {
```
Add whitespace after the colon?

shavavo replied:
Done

sagangwee commented on core/templates/domain/community_dashboard/FeaturedTranslationLanguageObjectFactory.ts:
```diff
@@ -0,0 +1,52 @@
+@Injectable({
+  providedIn: 'root'
+})
+export class FeaturedTranslationLanguageObjectFactory {
```
`Object` in the name seems unnecessary here.

shavavo replied:
Done

sagangwee commented on core/templates/domain/community_dashboard/FeaturedTranslationLanguageObjectFactory.ts:
```diff
@@ -0,0 +1,52 @@
+import { Injectable } from '@angular/core';
+
+export interface FeaturedTranslationLanguageBackendDict {
+  'language_code': string;
+  'description': string;
```
Nit: Prefer specifying the interface properties without the quotes.

shavavo replied:
Updated for description, but if we do this for language_code we get `Identifier 'language_code' is not in camel case.eslintcamelcase`

sagangwee replied:
Interesting. I guess since the backend dict is intended to be used in the Python backend code, we need it to be in snake_case, and keeping the field names as strings avoids the linter warning in somewhat of a hacky way. Then, it's probably fine to leave it as strings.

shavavo replied:
Done

Commit:
increased coverage by fixing async tests [skip ci]

sagangwee commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -130,6 +160,12 @@ export class ContributionOpportunitiesBackendApiService {
     });
   }
+
+  fetchFeaturedTranslationLanguages(): Promise<Object> {
+    return new Promise((resolve, reject) => {
+      this._getFeaturedTranslationLanguages(resolve, reject);
+    });
+  }
```
Hmm, instead of this, I feel it would be simpler to convert this public method and the private helper method into an [async](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Async_await) function.

shavavo replied:
@sagangwee I think async goes against convention of the code base, I don't see any other API calls using async/await

sagangwee replied:
That's probably due to the codebase being old, so most of the API usages are still using Promise chains (e.g. a lot of the TypeScript interfaces in the codebase are prefixed with 'I', but that's not necessarily a best practice). If you want to use a Promise chain here though, can't you just return the result of the `then()` call rather than explicitly invoking the success and failure callbacks? Then, you can inline the `fetchFeaturedTranslationLanguages` method instead of using a private helper method. 

At the very least, we would want to refactor the methods somehow so we don't have a method (`_getFeaturedTranslationLanguages` ) prefixed with an underscore.

shavavo replied:
Done

sagangwee reviewed (changes requested):
@shavavo Took another pass. Thanks!

Commit:
[skip ci] cleaned up code

shavavo commented:
@sagangwee PTAL

sagangwee reviewed (changes requested):
@shavavo Thanks! Just one minor comment. I'm definitely not an Angular/TypeScript expert, so if anyone else could chime in that would be great (@DubeySandeep, @marianazangrossi).

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

Commit:
async/await for backend

Commit:
try catch

Commit:
[skip ci] removed unnecessary try catch wrap

sagangwee commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -161,10 +138,18 @@ export class ContributionOpportunitiesBackendApiService {
-      this._getFeaturedTranslationLanguages(resolve, reject);
-    });
+  async fetchFeaturedTranslationLanguages(): Promise<Object> {
+    let response = await this.http
```
Use `const`?

shavavo replied:
Done

sagangwee reviewed (approved):
@shavavo Just one last comment, but otherwise LGTM. Thanks!

Commit:
let => const

shavavo commented:
@DubeySandeep PTAL

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

Commit:
fixed e2e test, backend lint, typescript eror

DubeySandeep commented on core/domain/config_domain.py:
```diff
@@ -30,6 +30,24 @@
+            'name': 'language_code',
+            'schema': {
+                'type': 'unicode'
+            },
```
should we add a validator `is_valid_translation_lanaguage_code`?

shavavo replied:
Done

DubeySandeep commented on core/templates/domain/community_dashboard/FeaturedTranslationLanguageFactory.ts:
```diff
@@ -0,0 +1,51 @@
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+export interface FeaturedTranslationLanguageBackendDict {
```
IFeaturedTranslationLanguageBackendDict (Interface has `I` prefix)

shavavo replied:
Done

DubeySandeep commented on core/domain/config_domain.py:
```diff
@@ -345,3 +363,10 @@ def get_all_config_property_names(cls):
     'classroom_page_is_shown', BOOL_SCHEMA,
     'Show classroom components.', False)
+
+FEATURED_TRANSLATION_LANGUAGES = ConfigProperty(
+    'featured_translation_languages',
+    LIST_OF_FEATURED_TRANSLATION_LANGUAGES_DICTS_SCHEMA,
+    'Featured Translation Languages',
+    []
+)
```
```suggestion
FEATURED_TRANSLATION_LANGUAGES = ConfigProperty(
    'featured_translation_languages',
    LIST_OF_FEATURED_TRANSLATION_LANGUAGES_DICTS_SCHEMA,
    'Featured Translation Languages', [])
```

shavavo replied:
Done

DubeySandeep commented on core/templates/domain/community_dashboard/FeaturedTranslationLanguageFactorySpec.ts:
```diff
@@ -0,0 +1,45 @@
+import { TestBed } from '@angular/core/testing';
+
+import { FeaturedTranslationLanguageFactory } from
+  './FeaturedTranslationLanguageFactory';
```
@bansalnitish, Can you confirm whether we follow this relative pattern for import?

shavavo replied:
Changed to absolute - Done

DubeySandeep commented on core/templates/domain/community_dashboard/FeaturedTranslationLanguageFactorySpec.ts:
```diff
@@ -0,0 +1,45 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
```
Rename the file to featured-translation-language.factory.spec.ts

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/community-dashboard-page.component.html:
```diff
@@ -52,14 +52,14 @@
-                ng-options="language.id as language.description for language in $ctrl.languageCodesAndDescriptions">
-        </select>
+        <div class="oppia-dashboard-language-container" ng-if="$ctrl.showLanguageSelector()">
+          <span class="oppia-dashboard-language-container-label">Translate to</span>
```
is it expected to remove `text` from `Translate text to:`?

shavavo replied:
Yes, from the mockups

DubeySandeep commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.spec.ts:
```diff
@@ -29,6 +29,10 @@ import { SkillOpportunityObjectFactory } from
   'domain/utilities/url-interpolation.service';
+import {
+  FeaturedTranslationLanguageFactory
+} from
```
Should we follow the existing pattern
```
import { ... } from
   ....
```

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -129,6 +136,20 @@ export class ContributionOpportunitiesBackendApiService {
   }
+
+  async fetchFeaturedTranslationLanguages(): Promise<Object> {
+    const response = await this.http
```
@bansalnitish, do we follow async/await pattern for the services?

shavavo replied:
Note - @sagangwee suggested that we start using async/await over promise chaining above

DubeySandeep replied:
@bansalnitish, PTAL!

DubeySandeep replied:
@bansalnitish, Can you please reply to this comment?

bansalnitish replied:
No, for the services I migrated or reviewed I didn't see this. What are your thoughts @srijanreddy98 ?

srijanreddy98 replied:
@DubeySandeep, @bansalnitish, and @sagangwee   We haven't used async-await in any of the services migrated yet. We don't have that sort of a pattern in services.
I think it would be better if we didn't introduce such a change now, as many of the services have been migrated and we would want to keep a uniform coding style.
That said, I am not against the idea of using async-await. In fact, I would prefer to have them over `promise.then()` in the directives/ components when we start migrating them. If it is used, then it should be done properly. The first thing to keep in mind is that async functions can only return promises. Secondly, there should be a catch for each await we use. Something like:
```
aysnc f() {
const result = await someAsyncTask().catch(err => {
  ...
});
```
or one could simply use try...catch.

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-select/translation-language-select.component.html:
```diff
@@ -0,0 +1,113 @@
+         *ngFor="let option of options"
+         (click)="selectOption(option.id)"
+         [ngClass]="{'oppia-translation-language-select-dropdown-option-selected' : option.id === value}"
+    >
```
Should we move this to the line above?

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-select/translation-language-select.component.html:
```diff
@@ -0,0 +1,113 @@
+
+<style>
+  .oppia-translation-language-select-container {
+    background-color: #ffffff;
```
#fff

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-select/translation-language-select.component.html:
```diff
@@ -0,0 +1,113 @@
+    display: flex;
+    justify-content: space-between;
+    padding: 10px 20px;
+    position: relative;
```
is this required for the child dom?

shavavo replied:
This is required to get child position: absolute elements to align correctly

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-select/translation-language-select.component.html:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  .oppia-translation-language-select-dropdown-container {
+    background-color: #ffffff;
```
CHange it to #fff (ditto elsewhere in this file)

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-select/translation-language-select.component.html:
```diff
@@ -0,0 +1,113 @@
+  .oppia-translation-language-select-container {
+    background-color: #ffffff;
+    border-radius: 4px;
+    box-shadow: 0px 0px 4px rgba(0, 0, 0, 0.25);
```
Don't use px for 0

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-select/translation-language-select.component.html:
```diff
@@ -0,0 +1,113 @@
+  .oppia-translation-language-select-dropdown-container {
+    background-color: #ffffff;
+    border-radius: 4px;
+    box-shadow: 0px 0px 4px rgba(0, 0, 0, 0.25);
```
Remove px from 0 (ditto elsewhere in this file

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-select/translation-language-select.component.spec.ts:
```diff
@@ -0,0 +1,192 @@
+  'domain/community_dashboard/FeaturedTranslationLanguageFactory';
+
+/**
+ * @fileoverview Unit tests for the translation language select
```
Add period `.` at the end.

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-select/translation-language-select.component.spec.ts:
```diff
@@ -0,0 +1,192 @@
+// limitations under the License.
+
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+import { TranslationLanguageSelectComponent } from
```
Add a newline to separate local and angular imports

shavavo replied:
Done

DubeySandeep commented on .github/CODEOWNERS:
```diff
@@ -70,6 +70,7 @@
 /core/domain/voiceover_services*.py @sagangwee
 /core/storage/opportunity/ @sagangwee
+/core/templates/domain/community_dashboard/ @sagangwee
 /core/templates/domain/opportunity/ @sagangwee
```
Should we move the new domain object file to this opportunity as opportunity sounds like a domain class and the new class is also related to opportunity. Any thoughts?

shavavo replied:
Done

DubeySandeep reviewed (changes requested):
@shavavo, Thanks for the PR I've left a few comments PTAL!

Commit:
Addressed review comments

Commit:
added validator

shavavo commented:
@DubeySandeep PTAL

shavavo commented:
Note: e2e failing because community dashboard was disabled, resulting in a 404

Commit:
fixed test

DubeySandeep commented:
@shavavo, I found the error log for community dashboard page a bit different ([link](https://travis-ci.com/github/oppia/oppia/jobs/350131858#L1796)). Can you reverify whether it's failing with 404 or something related to this PR? [AFAIK it's passing on develop so 404 is not expected!]

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

shavavo commented:
@DubeySandeep sorry about that, I wasn't running the tests with the flag ```--community_dashboard_enabled```

Commit:
fixed e2e test

shavavo commented:
@seanlip @DubeySandeep @bansalnitish PTAL as code owners. Note: lint tests failing due to issue #9458

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

seanlip commented on core/controllers/community_dashboard_test.py:
```diff
@@ -452,3 +453,29 @@ def test_user_check_community_rights(self):
+        new_value = [
+            {'language_code': 'en', 'description': 'Partnership with ABC'}
+        ]
+        config_services.set_property(
```
Perhaps also test setting the property to a value in an invalid format?

shavavo replied:
Good catch, though it probably doesn't belong in community_dashboard_test.py since they are not set there. I've added a test for the validator in schema_utils_test.py

seanlip commented on core/templates/domain/opportunity/featured-translation-language.factory.spec.ts:
```diff
@@ -0,0 +1,46 @@
+import { TestBed } from '@angular/core/testing';
+
+import { FeaturedTranslationLanguageFactory } from
+  // eslint-disable-next-line max-len
```
Do you actually need this? The line looks shorter than the one after it, to me.

shavavo replied:
Done

seanlip commented on core/templates/pages/community-dashboard-page/community-dashboard-page.component.html:
```diff
@@ -52,14 +52,14 @@
-        </select>
+        <div class="oppia-dashboard-language-container" ng-if="$ctrl.showLanguageSelector()">
+          <span class="oppia-dashboard-language-container-label">Translate to</span>
+          <translation-language-select [options]="$ctrl.languageCodesAndDescriptions"
```
Should be translation-language-selector perhaps, since that's more of a noun?

shavavo replied:
Done

seanlip commented on feconf.py:
```diff
@@ -434,7 +434,7 @@ def get_empty_ratings():
 
 # Whether community dashboard is ready to use for contributors.
-COMMUNITY_DASHBOARD_ENABLED = False
+COMMUNITY_DASHBOARD_ENABLED = True
```
Don't enable this yet please. (Right, @DubeySandeep?)

DubeySandeep replied:
Yeah, that's not expected! (We are going to enable it later once we have the images in suggestion functionality.)

shavavo replied:
sorry about that, enabled to test but forgot to disable

shavavo replied:
Done

seanlip reviewed (changes requested):
Only minor comments. Thanks @shavavo!

Commit:
addressed review comments

Commit:
Styling

Commit:
styling

Commit:
html styling

shavavo commented:
@seanlip PTAL

Commit:
fixed relative import

Commit:
fixed import

shavavo commented:
@DubeySandeep @bansalnitish @aks681 @kevintab95 PTAL as code owners.

seanlip commented on core/templates/pages/exploration-editor-page/translation-tab/translator-overview/translator-overview.directive.html:
```diff
@@ -57,7 +57,7 @@
     width: 350px;
   }
-  .oppia-translation-language-selector {
+  .oppia-translation-language-selectoror {
```
Copy-paste mistake, here and in a bunch of places elsewhere. 

Try to **never** copy-paste large-scale changes like this -- instead, make the changes manually and use grep to see if you have missed anything. Also, carefully proofread your changes as soon as (or ideally before) you submit them to catch things like this. You must **always** conduct a human review of each changed line somewhere in the process.

shavavo replied:
Yikes - my bad

shavavo replied:
Done

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

Commit:
typos

seanlip reviewed (approved):
LGTM. Thanks @shavavo!

seanlip commented:
Btw -- I am a bit confused, for this PR, about why @aks681 @kevintab95 started off as codeowners and then stopped being codeowners, when (AFAICT) nothing substantial changed. Any idea why?

Also @bansalnitish @DubeySandeep PTAL, thanks!

kevintab95 commented:
Looks like 39ce227 reverted an unnecessary change in 2 files. @aks681 and I were codeowners for both files. So now that there are no changes in those files in this PR, codeowner reviews are not required from us.

seanlip commented:
Ah! I see, thanks so much for following up on that @kevintab95! Good to know.

bansalnitish commented on core/templates/services/UpgradedServices.ts:
```diff
@@ -164,6 +164,8 @@ import { ExtensionTagAssemblerService } from
 import { ExtractImageFilenamesFromStateService } from
   // eslint-disable-next-line max-len
   'pages/exploration-player-page/services/extract-image-filenames-from-state.service';
+import { FeaturedTranslationLanguageFactory } from
```
This needs to be added in the `oppia-angular-root.component.ts ` as well, right @srijanreddy98 ?

Please follow point number 12 under this: https://github.com/oppia/oppia/wiki/Angular-Migration#implementation-details-to-migrate-services.

shavavo replied:
Done

kevintab95 commented:
@DubeySandeep code-owner review please!

DubeySandeep commented on core/templates/domain/opportunity/FeaturedTranslationLanguageFactory.ts:
```diff
@@ -0,0 +1,51 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
Should we also rename this file with the same pattern?

shavavo replied:
Done

DubeySandeep commented on core/templates/domain/opportunity/FeaturedTranslationLanguageFactory.ts:
```diff
@@ -0,0 +1,51 @@
+export class FeaturedTranslationLanguage {
+  constructor(
+      readonly languageCode: string,
+      readonly description: string
```
Is this going to be the same `description` as the description of a language in `SUPPORTED_AUDIO_LANGUAGES`? If not, then I think this can be confusing later on. Can we call it `annotation` or `explnantion` (or anything around that)?

shavavo replied:
Done

DubeySandeep commented on core/templates/domain/opportunity/FeaturedTranslationLanguageFactory.ts:
```diff
@@ -0,0 +1,51 @@
+export class FeaturedTranslationLanguageFactory {
+  createFromBackendDict(
+      featuredTranslationBackendDict:IFeaturedTranslationLanguageBackendDict
+  ) : FeaturedTranslationLanguage {
```
Do we need the space after `)`?

shavavo replied:
Done

DubeySandeep commented on core/templates/domain/opportunity/FeaturedTranslationLanguageFactory.ts:
```diff
@@ -0,0 +1,51 @@
+})
+export class FeaturedTranslationLanguageFactory {
+  createFromBackendDict(
+      featuredTranslationBackendDict:IFeaturedTranslationLanguageBackendDict
```
Should we add a space after `:`?

shavavo replied:
Done

DubeySandeep commented on core/templates/domain/opportunity/featured-translation-language.factory.spec.ts:
```diff
@@ -0,0 +1,45 @@
+describe('Featured Translation Language object factory', () => {
+  let featuredTranslationLanguageFactory:
+    FeaturedTranslationLanguageFactory = null;
+  let sampleFTL = null;
```
should be add `IFeaturedTranslationLanguageBackendDict` type here? (not sure!)

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/community-dashboard-page.component.html:
```diff
@@ -80,23 +80,22 @@
-    width: 15%;
+    width: 60%;
   }
+
```
extra newline?

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/community-dashboard-page.component.html:
```diff
@@ -52,14 +52,14 @@
-        </select>
+        <div class="oppia-dashboard-language-container" ng-if="$ctrl.showLanguageSelector()">
+          <span class="oppia-dashboard-language-container-label">Translate to</span>
+          <translation-language-selector [options]="$ctrl.languageCodesAndDescriptions"
```
Why do we need to pass these values to the directive I think the directive can self generate this isn't it? (Considering the  name of the directive contains `translation language` so is should know all the language-related information.)

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.html:
```diff
@@ -0,0 +1,112 @@
+<div #dropdown class="oppia-translation-language-selector-container protractor-test-language-selector">
+  <div class="oppia-translation-language-selector-inner-container" (click)="toggleDropdown()">
+    {{languageIdToDescription[value]}}
+    <span class="material-icons" style="color: #009688;" [ngStyle]="{ 'transform': dropdownShown ? 'scaleY(-1)' : 'scaleY(1)' }">
```
Do not use inline style here! `style="color: #009688;"`

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -129,6 +135,20 @@ export class ContributionOpportunitiesBackendApiService {
+
+  async fetchFeaturedTranslationLanguages(): Promise<Object> {
+    const response = await this.http
+      .get('/getfeaturedtranslationlanguages')
```
'get' sounds like a typical getter, which this isn't, should we change it to `retrivefeaturedtranslationlanguages`?

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -0,0 +1,101 @@
+        .then((data: FeaturedTranslationLanguage[]) => {
+          this.featuredLanguages = data;
+        });
+    } catch {}
```
Any specific reason for adding this in a try catch block? (new to angular components.)

shavavo replied:
If the http call fails, the catch makes sure we don't throw an error

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -0,0 +1,101 @@
+      this.contributionOpportunitiesBackendApiService
+        .fetchFeaturedTranslationLanguages()
+        .then((data: FeaturedTranslationLanguage[]) => {
+          this.featuredLanguages = data;
```
Can we have a better name instead of data? as that's a domain objects, maybe featuredLanguages?

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  showDescriptionPopup(index: number) {
+    this.descriptionPopupPxOffsetY = 75 + 30 * index;
```
Can you explain these random numbers? (Add a code comment on top.)

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -0,0 +1,101 @@
+    const targetElement = event.target as HTMLElement;
+    if (
+      targetElement &&
+      !this.dropdownRef.nativeElement.contains(targetElement)
```
Can you explain what it does and when this event gets triggered?

shavavo replied:
Done

DubeySandeep commented on core/tests/protractor_utils/CommunityDashboardTranslateTextTab.js:
```diff
@@ -23,11 +23,15 @@ var CommunityDashboardTranslateTextTab = function() {
+    by.css('.oppia-translation-language-selector-inner-container'));
 
   var _selectLanguage = async function(language) {
+    await selectableLanguageElements.click();
```
Add `await for element to be clickable` before this.

shavavo replied:
Done

DubeySandeep commented on core/tests/protractor_utils/CommunityDashboardTranslateTextTab.js:
```diff
@@ -23,11 +23,15 @@ var CommunityDashboardTranslateTextTab = function() {
     by.css('.protractor-test-language-selector'));
   var selectedLanguageElement = selectableLanguageElements.element(
-    by.css('option:checked'));
+    by.css('.oppia-translation-language-selector-inner-container'));
```
Do not use `oppia-` classes here, create a separate `protractor-test-` prefix class.

shavavo replied:
Done

DubeySandeep commented on core/tests/protractor_utils/CommunityDashboardTranslateTextTab.js:
```diff
@@ -23,11 +23,15 @@ var CommunityDashboardTranslateTextTab = function() {
     await selectableLanguageElements.element(
-      by.cssContainingText('option', language)).click();
+      by.cssContainingText(
+        '.oppia-translation-language-selector-dropdown-option',
```
Ditto don't use `oppia-` prefix class.

shavavo replied:
Done

DubeySandeep commented on core/tests/protractor_utils/CommunityDashboardTranslateTextTab.js:
```diff
@@ -23,11 +23,15 @@ var CommunityDashboardTranslateTextTab = function() {
+      by.cssContainingText(
+        '.oppia-translation-language-selector-dropdown-option',
+        language
+      )).click();
```
Add e2e test for the feature language explanation text.

DubeySandeep commented on main.py:
```diff
@@ -265,6 +265,9 @@ def ui_access_wrapper(self, *args, **kwargs):
         r'/usercommunityrightsdatahandler',
         community_dashboard.UserCommunityRightsDataHandler),
+    get_redirect_route(
+        r'/getfeaturedtranslationlanguages',
```
I've left a comment on top for the name of this url, we can change it to the name of the controller  `featuredtranslationlanguageshandler`, what do you think?

shavavo replied:
Done

DubeySandeep commented on schema_utils.py:
```diff
@@ -447,3 +448,15 @@ def is_valid_math_equation(obj):
+        Returns:
+            bool. Whether the given object is a valid language code.
+        """
+        return utils.is_valid_language_code(obj)
```
No, it should validate whether that's a valid translation language code. Maybe `utils.is_supported_audio_language_code`? (Note as of now the audio and translation languages are the same set of languages)

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -0,0 +1,195 @@
+ * @fileoverview Unit tests for the translation language select.
+ */
+
+describe('Translation language select', () => {
```
selector*

shavavo replied:
Done

DubeySandeep reviewed (changes requested):
Hi @shavavo, I've left a few comments PTAL! Also, can we move the selector to the same position it was earlier? Currently, it looks more packed in one place.

DubeySandeep commented:
![image](https://user-images.githubusercontent.com/16653571/85232262-54520700-b41b-11ea-97cc-6fbc9745007d.png)
Should we change the scrollbar style to follow the other scroll bars on the dashboard page? You can check `pages/community-dashboard-page/contributions-and-review/contributions-and-review.directive.html` file.

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

Commit:
review comments

Commit:
syntax

Commit:
syntax

Commit:
tests

shavavo commented:
@DubeySandeep PTAL. I moved it back to the original position, and updated the scrollbar. Note e2e tests may be flaky right now, will investigate further tomorrow. Thanks!

Commit:
styling

DubeySandeep commented on core/tests/protractor_desktop/communityDashboard.js:
```diff
@@ -17,7 +17,9 @@
 var general = require('../protractor_utils/general.js');
-
+var users = require('../protractor_utils/users.js');
+var waitFor = require('../protractor_utils/waitFor.js');
```
Add a newline below.

shavavo replied:
Done

Commit:
MC

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

DubeySandeep commented on core/templates/domain/opportunity/featured-translation-language-object.factory.spec.ts:
```diff
@@ -0,0 +1,49 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Tests for ReadOnlyStoryNodeObjectFactory.
```
Add correct factory name

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.spec.ts:
```diff
@@ -43,7 +45,7 @@ describe('Contribution Opportunities backend API service', function() {
     opportunities: [{
       id: 'skill_id',
-      skill_description: 'A new skill for question',
+      skill_explanation: 'A new skill for question',
```
is this an expected change?

shavavo replied:
No - sorry about that. Fixed

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -0,0 +1,162 @@
+  'domain/opportunity/FeaturedTranslationLanguageObjectFactory';
+
+/**
+ * @fileoverview Unit tests for the translation language select.
```
selector component*

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -0,0 +1,114 @@
+  templateUrl: './translation-language-selector.component.html'
+})
+export class TranslationLanguageSelectorComponent implements OnInit {
+  @Input() value: string;
```
should we call this activeLanguageCode? Ditto below maybe setActiveLanguage?

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  showExplanationPopup(index: number) {
+    // Align popup to mouse-overed info icon
```
End comment with a period*.

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -0,0 +1,114 @@
+
+  showExplanationPopup(index: number) {
+    // Align popup to mouse-overed info icon
+    this.explanationPopupPxOffsetY = 75 + 30 * index;
```
Is there any specific calculation behind 75 and 30, if yes can you please add info through a code comment?

shavavo replied:
There isnt any specific calculation, they are approximate, but I added a comment description for the numbers

nishantwrp commented:
@shavavo please note the failing ci checks

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -0,0 +1,114 @@
+   * Close popup when outside elements are clicked
+   * @param event mouse click event
+   */
+  @HostListener('document:click', ['$event'])
```
Should we scope this event listener to small dom? Also, have you tried using tooltip or popover (see line 270 of editor-navigation.directive), can we do it that way?

shavavo replied:
I believe the click listener must be attached to the document, correct me if I'm wrong. Also, I've looked into uib-tooltip, but I'm not sure that we have bootstrap installed for Angular2+, @marianazangrossi could you weigh in? 

Otherwise, the tooltip I made works fine, I believe.

marianadasilvadev replied:
Hi @shavavo, I think you want to ask for @bansalnitish and @srijanreddy98, who are responsible for Angular 2+ features.

shavavo replied:
Update: from email thread with Nitish and Srijan, adding bootstrap + material design is currently being investigated

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -0,0 +1,114 @@
+   */
+  @HostListener('document:click', ['$event'])
+  onDocumentClick(event: MouseEvent): void {
+    const targetElement = event.target as HTMLElement;
```
Also, should we only show the explanation until the user hovers over the `i` button? (not sure!)

shavavo replied:
Yes, that is what is happening currently. onDocumentClick closes the dropdown

DubeySandeep commented on core/tests/protractor_utils/CommunityDashboardTranslateTextTab.js:
```diff
@@ -20,21 +20,75 @@
+    );
+    await browser.actions().mouseMove(
+      featuredLanguageTooltipElements.get(index)
+    ).perform();
```
Do we need two `await` here? One for mouseMove and others for perform? (not sure!)

shavavo replied:
No, I don't believe so

DubeySandeep commented on core/tests/protractor_utils/CommunityDashboardTranslateTextTab.js:
```diff
@@ -20,21 +20,75 @@
+  };
+
+  this.expectFeaturedLanguagesToBe = async function(languages) {
+    await _openSelector();
```
_openLanguageSelector

shavavo replied:
Done

DubeySandeep commented on schema_utils.py:
```diff
@@ -447,3 +448,15 @@ def is_valid_math_equation(obj):
         return False
+
+    @staticmethod
+    def is_supported_audio_language_code(obj):
```
change obj to languageCode

shavavo replied:
Done

shavavo replied:
@DubeySandeep this change was reverted back to `obj`. `language_code` was causing a validation test to fail (```ConfigPropertyRegistryTests.test_config_property_schemas_are_valid```)

DubeySandeep replied:
I see, thanks for the info!

DubeySandeep reviewed (changes requested):
LGTM  just fee minor comments PTAL!

Commit:
remove fit

Commit:
review

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

Commit:
doc string

Commit:
test

Commit:
conf

Commit:
fixed frontend tests

shavavo commented:
@DubeySandeep @bansalnitish  @srijanreddy98  PTAL

DubeySandeep commented:
@shavavo, I think you missed [this comment](https://github.com/oppia/oppia/pull/9443#discussion_r444990299), PTAL!

shavavo commented:
@DubeySandeep Mb, addressed your comment. PTAL

Commit:
fixed test

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

marianadasilvadev commented on core/templates/domain/opportunity/featured-translation-language-object.factory.spec.ts:
```diff
@@ -0,0 +1,49 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
```
Change the current year.

shavavo replied:
Done

marianadasilvadev commented on core/templates/domain/opportunity/featured-translation-language-object.factory.spec.ts:
```diff
@@ -0,0 +1,49 @@
+  });
+
+  it('should correctly return all the values', function() {
+    expect(sampleFTL.languageCode).toEqual('en');
+    expect(sampleFTL.explanation).toEqual('English');
```
```suggestion
    expect(sampleFTL.languageCode).toBe('en');
    expect(sampleFTL.explanation).toBe('English');
```

shavavo replied:
Done

marianadasilvadev commented on core/templates/domain/opportunity/featured-translation-language-object.factory.spec.ts:
```diff
@@ -0,0 +1,49 @@
+      .createFromBackendDict(sampleFTLDict);
+  });
+
+  it('should correctly return all the values', function() {
```
```suggestion
  it('should correctly evaluate all the values based on backend dict', function() {
```

shavavo replied:
Done

marianadasilvadev commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -0,0 +1,162 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { TranslationLanguageSelectorComponent } from
+  // eslint-disable-next-line max-len
+  'pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.ts';
+import { ContributionOpportunitiesBackendApiService } from
+  // eslint-disable-next-line max-len
+  'pages/community-dashboard-page/services/contribution-opportunities-backend-api.service.ts';
+import { FeaturedTranslationLanguageObjectFactory } from
+  'domain/opportunity/FeaturedTranslationLanguageObjectFactory';
```
Change all these imports to be after fileoverview line.

shavavo replied:
Done

marianadasilvadev commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -0,0 +1,162 @@
+    expect(getDropdownOptionsContainer()).toBeFalsy();
+  });
+
+
```
Remove new line

shavavo replied:
Done

marianadasilvadev commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -0,0 +1,162 @@
+        provide: ContributionOpportunitiesBackendApiService,
+        useValue: contributionOpportunitiesBackendApiServiceStub
+      }]
+    })
+      .compileComponents();
```
```suggestion
    }).compileComponents();
```

shavavo replied:
Done

marianadasilvadev commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -0,0 +1,162 @@
+  });
+
+  it('should correctly initialize languageIdToDescription map', () => {
+    expect(component.languageIdToDescription.en).toEqual('English');
+    expect(component.languageIdToDescription.fr).toEqual('French');
```
```suggestion
    expect(component.languageIdToDescription.en).toBe('English');
    expect(component.languageIdToDescription.fr).toBe('French');
```

shavavo replied:
Done

marianadasilvadev commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -0,0 +1,162 @@
+    const dropdown = (fixture.nativeElement
+      .querySelector('.oppia-translation-language-selector-inner-container'));
+
+    expect(dropdown.firstChild.textContent.trim()).toEqual('English');
```
```suggestion
    expect(dropdown.firstChild.textContent.trim()).toBe('English');
```

Here and elsewhere. Just to clarify, the `toBe` method is to check primitive values as number, string, boolean. The `toEqual` method checks reference values, as object and arrays,

shavavo replied:
Done

marianadasilvadev reviewed (changes requested):
I'm a reviewer after commenting out here, so I've taken a pass through some unit tests.

Commit:
styling

shavavo commented:
@marianazangrossi Thanks, Mariana. PTAL

shavavo commented:
@DubeySandeep @bansalnitish @srijanreddy98 PTAL

Commit:
fixed test

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

shavavo commented:
Added a TODO: #9674 for @shavavo after merge

marianadasilvadev commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -0,0 +1,160 @@
+  let component: TranslationLanguageSelectorComponent;
+  let fixture: ComponentFixture<TranslationLanguageSelectorComponent>;
+
+  let featuredTranslationLanguageObjectFactory =
+    new FeaturedTranslationLanguageObjectFactory();
```
```suggestion
  let featuredTranslationLanguageObjectFactory = (
    new FeaturedTranslationLanguageObjectFactory());
```

shavavo replied:
Done

marianadasilvadev reviewed (approved):
LGTM, thanks @shavavo! Just left a minor nit.

shavavo committed:
Update core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts

Co-authored-by: Mariana Zangrossi <34922478+marianazangrossi@users.noreply.github.com>

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.html:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  .oppia-translation-language-selector-featured-tooltip .oppia-translation-language-selector-featured-tooltip-content {
+    background-color: #ffffff;
```
use #fff

shavavo replied:
Done

DubeySandeep commented on core/templates/pages/community-dashboard-page/translation-language-selector/translation-language-selector.component.html:
```diff
@@ -0,0 +1,135 @@
+    box-shadow: inset 0 0 3px grey;
+  }
+  .oppia-translation-language-selector-dropdown-container::-webkit-scrollbar-thumb {
+    background: grey;
```
Use hex value (also instead of rgb(..., ..., ...) can we use hex for consistency?)

shavavo replied:
Done

Commit:
async try catch

DubeySandeep reviewed (commented):
Resolve the comments I've left below + two other existing comments and we can merge it! :)

Commit:
Merge branch 'translation-prioritization' of github.com:shavavo/oppia into translation-prioritization

Commit:
styling

bansalnitish reviewed (approved):
LGTM for codeowners file.

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

DubeySandeep reviewed (approved):
LGTM!

Commit:
Merge remote-tracking branch 'upstream/develop' into translation-prioritization

shavavo commented:
@srijanreddy98  PTAL

srijanreddy98 reviewed (approved):
LGTM!
